spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/ConvertingMessageListener.java

```diff
@@ -0,0 +1,127 @@
+import org.springframework.util.Assert;
+
+/**
+ * A {@link AcknowledgingConsumerAwareMessageListener} adapter that implements
+ * converting received {@link ConsumerRecord} using specified {@link MessageConverter}
+ * and then passes result to specified {@link MessageListener}.
+ *
+ * @param <V> the desired value type after conversion.
+ *
+ * @author Adrian Chlebosz
+ * @since 3.0
+ * @see AcknowledgingConsumerAwareMessageListener
+ */
+@SuppressWarnings("rawtypes")
+public class ConvertingMessageListener<V> implements AcknowledgingConsumerAwareMessageListener<Object, Object> {
+
+	private final MessageListener delegate;
+	private final Class<V> desiredValueType;
+
+	private MessageConverter messageConverter;
+
+	/**
+	 * Construct an instance with the provided {@link MessageListener} and {@link Class}
+	 * as a desired type of {@link ConsumerRecord}'s value after conversion. Default value of
+	 * {@link MessageConverter} is used, which is {@link GenericMessageConverter}.
+	 *
+	 * @param delegateMessageListener the {@link MessageListener} to use when passing converted {@link ConsumerRecord} further.
+	 * @param desiredValueType the {@link Class} setting desired type of {@link ConsumerRecord}'s value.
+	 */
+	public ConvertingMessageListener(MessageListener<?, V> delegateMessageListener, Class<V> desiredValueType) {
+		Assert.notNull(delegateMessageListener, "'delegateMessageListener' cannot be null");
+		Assert.notNull(desiredValueType, "'desiredValueType' cannot be null");
+		this.delegate = delegateMessageListener;
+		this.desiredValueType = desiredValueType;
+
+		this.messageConverter = new GenericMessageConverter();
+	}
+
+	/**
+	 * Set a {@link MessageConverter}.
+	 * @param messageConverter the message converter to use for conversion of incoming {@link ConsumerRecord}.
+	 * @since 3.0
+	 */
+	public void setMessageConverter(MessageConverter messageConverter) {
+		Assert.notNull(messageConverter, "'messageConverter' cannot be null");
+		this.messageConverter = messageConverter;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void onMessage(ConsumerRecord data, Acknowledgment acknowledgment, Consumer consumer) {
+		ConsumerRecord convertedConsumerRecord = convertConsumerRecord(data);
+		if (this.delegate instanceof AcknowledgingConsumerAwareMessageListener) {
+			this.delegate.onMessage(convertedConsumerRecord, acknowledgment, consumer);
+		}
+		else if (this.delegate instanceof ConsumerAwareMessageListener) {
+			this.delegate.onMessage(convertedConsumerRecord, consumer);
+		}
+		else if (this.delegate instanceof AcknowledgingMessageListener) {
+			this.delegate.onMessage(convertedConsumerRecord, acknowledgment);
+		}
+
+		this.delegate.onMessage(convertedConsumerRecord);
+	}
+
+	private ConsumerRecord convertConsumerRecord(ConsumerRecord data) {
+		Header[] headerArray = data.headers().toArray();
+		Map<String, Object> headerMap = Arrays.stream(headerArray)
+				.collect(Collectors.toMap(Header::key, Header::value));
```

Comment:
I analysed the code one more time and figured out that creating GenericMessage with headers is pointless, because they don't play any role in conversion of message payload. Also in newly built ConsumerRecord they are just copied form received record. Based on that I deleted these lines related to extracting headers and take care only about payload in conversion. What's your opinion?